src/modules/schnorr_adaptor/main_impl.h

```diff
@@ -0,0 +1,286 @@
+        secp256k1_nonce_function_bip340_sha256_tagged(&sha);
+    } else {
+        secp256k1_sha256_initialize_tagged(&sha, algo, algolen);
+    }
+
+    /* Hash masked-key||pk||msg using the tagged hash as per the spec */
+    secp256k1_sha256_write(&sha, masked_key, 32);
+    secp256k1_sha256_write(&sha, xonly_t32, 32);
+    secp256k1_sha256_write(&sha, xonly_pk32, 32);
+    secp256k1_sha256_write(&sha, msg32, 32);
+    secp256k1_sha256_finalize(&sha, nonce32);
+    return 1;
+}
+
+const secp256k1_adaptor_nonce_function_hardened secp256k1_adaptor_nonce_function_bip340 = adaptor_nonce_function_bip340;
+
+static int secp256k1_schnorr_adaptor_presign_internal(const secp256k1_context *ctx, unsigned char *sig65, const unsigned char *msg32, const secp256k1_keypair *keypair, secp256k1_adaptor_nonce_function_hardened noncefp, const unsigned char *t33, void *ndata) {
+    secp256k1_scalar sk;
+    secp256k1_scalar e;
+    secp256k1_scalar k;
+    secp256k1_gej rj;
+    secp256k1_gej r0j;
+    secp256k1_ge pk;
+    secp256k1_ge r;
+    secp256k1_ge r0;
+    secp256k1_ge t;
+    unsigned char nonce32[32] = {0};
+    unsigned char pk_buf[32];
+    unsigned char seckey[32];
+    size_t size = 33;
+    size_t msglen = 32;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));
+    ARG_CHECK(sig65 != NULL);
+    ARG_CHECK(msg32 != NULL);
+    ARG_CHECK(keypair != NULL);
+    ARG_CHECK(t33 != NULL);
+
+    if (noncefp == NULL) {
+        noncefp = secp256k1_adaptor_nonce_function_bip340;
+    }
+
+    ret &= secp256k1_keypair_load(ctx, &sk, &pk, keypair);
+
+    if (secp256k1_fe_is_odd(&pk.y)) {
+        secp256k1_scalar_negate(&sk, &sk);
+    }
+
+    /* d */
+    secp256k1_scalar_get_b32(seckey, &sk);
+    /* bytes_from_point(P) */
+    secp256k1_fe_get_b32(pk_buf, &pk.x);
+
+    ret &= !!noncefp(nonce32, msg32, seckey, &t33[1], pk_buf, bip340_algo, sizeof(bip340_algo), ndata);
+    /* k0 */
+    secp256k1_scalar_set_b32(&k, nonce32, NULL);
+    ret &= !secp256k1_scalar_is_zero(&k);
+    secp256k1_scalar_cmov(&k, &secp256k1_scalar_one, !ret);
+
+    /* R = k0*G */
+    secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &rj, &k);
+    secp256k1_ge_set_gej(&r, &rj);
+
+    /* T = cpoint(T) */
+    ret &= !!secp256k1_eckey_pubkey_parse(&t, t33, 33);
+
+    /* R' = k*G + T, can use gej_add_ge_var since r and t aren't secret */
+    secp256k1_gej_add_ge_var(&r0j, &rj, &t, NULL);
+    secp256k1_ge_set_gej(&r0, &r0j);
+
+    /* We declassify R' to allow using it as a branch point. This is fine
+     * because R' is not a secret.  */
+    secp256k1_declassify(ctx, &r0, sizeof(r0));
+    secp256k1_fe_normalize_var(&r0.y);
+    if (secp256k1_fe_is_odd(&r0.y)) {
+        secp256k1_scalar_negate(&k, &k);
+    }
+
+    ret &= !!secp256k1_eckey_pubkey_serialize(&r0, sig65, &size, 1);
+
+    secp256k1_schnorrsig_challenge(&e, &sig65[1], msg32, msglen, pk_buf);
```

Comment:
The challenge function here should be consistent with the schnorrsig challenge function, it just hash the x coordinate of R', and that is okay since we have the parity reordered in the first half of the signature.